File: retaRust/python_arch_reference/reta_architecture/column_selection.py
from __future__ import annotations

from collections import namedtuple
from dataclasses import dataclass
from typing import Callable, Dict


COLUMN_BUCKET_NAMES: tuple[str, ...] = (
    "ordinary",
    "generated1",
    "concat1",
    "kombi1",
    "boolAndTupleSet1",
    "gebroUni1",
    "gebrGal1",
    "generated2",
    "kombi2",
    "gebrEmo1",
    "gebrGroe1",
    "metakonkret",
    "ordinaryNot",
    "generate1dNot",
    "concat1Not",
    "kombi1Not",
    "boolAndTupleSet1Not",
    "gebroUni1Not",
    "gebrGal1Not",
    "generated2Not",
    "kombi2Not",
    "gebrEmo1Not",
    "gebrGroe1Not",
    "metakonkretNot",
)

_COLUMN_TYPE = namedtuple("SpaltenTyp", " ".join(COLUMN_BUCKET_NAMES))
_COLUMN_BUCKET_VALUES: tuple[tuple[int, int], ...] = tuple(
    (negation, bucket)
    for negation in (0, 1)
    for bucket in range(12)
)


@dataclass(frozen=True)
class ColumnSelectionBundle:
    """Legacy-compatible column-bucket model for reta's Spalten selection.

    This is the explicit architecture layer for the old local pair-space
    `(positive|negative, bucket-type)` used by `reta.Program`. It keeps the
    exact field names and bucket coordinates, but stops `reta.py` from owning the
    schema inline.
    """

    ordered_set_factory: Callable[[], object] = set

    @property
    def type_naming(self):
        return _COLUMN_TYPE(*_COLUMN_BUCKET_VALUES)

    @property
    def bucket_values(self) -> tuple[tuple[int, int], ...]:
        return _COLUMN_BUCKET_VALUES

    @property
    def bucket_names(self) -> tuple[str, ...]:
        return COLUMN_BUCKET_NAMES

    def new_bucket_map(self) -> Dict[tuple[int, int], object]:
        return {key: self.ordered_set_factory() for key in self.bucket_values}

    def bind_program_sections(self, program, param_lines) -> None:
        """Bind normalized bucket sections back onto the legacy Program object.

        The method is deliberately side-effectful: it is a compatibility adapter
        for the current Program runtime. Architecturally, this is the point where
        the local column-bucket presheaf is interpreted as the program's concrete
        selected column sections.
        """

        buckets = program.spaltenArtenKey_SpaltennummernValue
        naming = program.spaltenTypeNaming

        program.rowsAsNumbers = buckets[naming.ordinary]
        program.generRows = buckets[naming.generated1]
        program.puniverseprims = buckets[naming.concat1]
        program.rowsOfcombi = buckets[naming.kombi1]
        program.rowsOfcombi2 = buckets[naming.kombi2]
        program.onlyGenerated = buckets[naming.boolAndTupleSet1]

        ones = []
        for item in program.onlyGenerated:
            try:
                if len(item) == 1:
                    ones += item
            except TypeError:
                continue
        program.tables.getConcat.ones = ones

        if len(program.rowsOfcombi) > 0:
            param_lines.add("ka")
        if len(program.rowsOfcombi2) > 0:
            param_lines.add("ka2")

        program.tables.generRows = program.generRows
        program.tables.getPrepare.rowsAsNumbers = program.rowsAsNumbers
        program.tables.getOut.rowsAsNumbers = program.rowsAsNumbers
        program.tables.SpaltenVanillaAmount = len(program.rowsAsNumbers)

    def snapshot(self) -> dict:
        return {
            "class": "ColumnSelectionBundle",
            "bucket_names": list(self.bucket_names),
            "bucket_values": [list(value) for value in self.bucket_values],
            "positive_bucket_count": 12,
            "negative_bucket_count": 12,
        }


def bootstrap_column_selection(ordered_set_factory=None) -> ColumnSelectionBundle:
    return ColumnSelectionBundle(ordered_set_factory=ordered_set_factory or set)

File: retaRust/crates/reta_architecture/src/column_selection.rs
//! Column-bucket selection layer transcompiled from
//! `python_arch_reference/reta_architecture/column_selection.py`.
//!
//! Reta's old `Program` uses a two-dimensional bucket space
//! `(positive|negative, bucket-type)`.  This module owns that schema in Rust so
//! the CLI parser, table generation and prompt completions can share the same
//! typed coordinates instead of re-creating Python namedtuple fields.

use std::collections::{BTreeMap, BTreeSet};

use serde::{Deserialize, Serialize};

pub const COLUMN_BUCKET_NAMES: [&str; 24] = [
    "ordinary",
    "generated1",
    "concat1",
    "kombi1",
    "boolAndTupleSet1",
    "gebroUni1",
    "gebrGal1",
    "generated2",
    "kombi2",
    "gebrEmo1",
    "gebrGroe1",
    "metakonkret",
    "ordinaryNot",
    "generate1dNot",
    "concat1Not",
    "kombi1Not",
    "boolAndTupleSet1Not",
    "gebroUni1Not",
    "gebrGal1Not",
    "generated2Not",
    "kombi2Not",
    "gebrEmo1Not",
    "gebrGroe1Not",
    "metakonkretNot",
];

pub const COLUMN_BUCKET_VALUES: [(u8, u8); 24] = [
    (0, 0),
    (0, 1),
    (0, 2),
    (0, 3),
    (0, 4),
    (0, 5),
    (0, 6),
    (0, 7),
    (0, 8),
    (0, 9),
    (0, 10),
    (0, 11),
    (1, 0),
    (1, 1),
    (1, 2),
    (1, 3),
    (1, 4),
    (1, 5),
    (1, 6),
    (1, 7),
    (1, 8),
    (1, 9),
    (1, 10),
    (1, 11),
];

#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
pub struct ColumnBucketKey {
    pub negation: u8,
    pub bucket: u8,
}

impl ColumnBucketKey {
    pub const fn new(negation: u8, bucket: u8) -> Self {
        Self { negation, bucket }
    }

    pub const fn positive(bucket: u8) -> Self {
        Self::new(0, bucket)
    }

    pub const fn negative(bucket: u8) -> Self {
        Self::new(1, bucket)
    }

    pub fn as_tuple(self) -> (u8, u8) {
        (self.negation, self.bucket)
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ColumnBucketSnapshot {
    pub name: String,
    pub negation: u8,
    pub bucket: u8,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ColumnSelectionSnapshot {
    pub class: String,
    pub bucket_names: Vec<String>,
    pub bucket_values: Vec<[u8; 2]>,
    pub positive_bucket_count: usize,
    pub negative_bucket_count: usize,
    pub universal_property: String,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct ColumnSelectionBundle;

impl ColumnSelectionBundle {
    pub fn bucket_names(&self) -> Vec<String> {
        COLUMN_BUCKET_NAMES.iter().map(|value| (*value).to_string()).collect()
    }

    pub fn bucket_values(&self) -> Vec<ColumnBucketKey> {
        COLUMN_BUCKET_VALUES
            .iter()
            .map(|(negation, bucket)| ColumnBucketKey::new(*negation, *bucket))
            .collect()
    }

    pub fn type_naming(&self) -> BTreeMap<String, ColumnBucketKey> {
        COLUMN_BUCKET_NAMES
            .iter()
            .zip(COLUMN_BUCKET_VALUES.iter())
            .map(|(name, (negation, bucket))| {
                ((*name).to_string(), ColumnBucketKey::new(*negation, *bucket))
            })
            .collect()
    }

    pub fn new_bucket_map(&self) -> BTreeMap<ColumnBucketKey, BTreeSet<i64>> {
        self.bucket_values()
            .into_iter()
            .map(|key| (key, BTreeSet::new()))
            .collect()
    }

    pub fn normalize_bucket_map(
        &self,
        buckets: &BTreeMap<ColumnBucketKey, BTreeSet<i64>>,
    ) -> BTreeMap<ColumnBucketKey, BTreeSet<i64>> {
        let mut out = buckets.clone();
        for bucket in 0..12u8 {
            let positive = ColumnBucketKey::positive(bucket);
            let negative = ColumnBucketKey::negative(bucket);
            if let Some(negative_values) = out.get(&negative).cloned() {
                if let Some(positive_values) = out.get_mut(&positive) {
                    for value in &negative_values {
                        positive_values.remove(value);
                    }
                }
                out.remove(&negative);
            }
        }
        out
    }

    pub fn snapshot(&self) -> ColumnSelectionSnapshot {
        ColumnSelectionSnapshot {
            class: "ColumnSelectionBundle".to_string(),
            bucket_names: self.bucket_names(),
            bucket_values: COLUMN_BUCKET_VALUES
                .iter()
                .map(|(negation, bucket)| [*negation, *bucket])
                .collect(),
            positive_bucket_count: 12,
            negative_bucket_count: 12,
            universal_property:
                "positive column buckets glue after subtracting matching negative local sections"
                    .to_string(),
        }
    }
}

pub fn bootstrap_column_selection() -> ColumnSelectionBundle {
    ColumnSelectionBundle
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn names_match_python_namedtuple_order() {
        let bundle = bootstrap_column_selection();
        let naming = bundle.type_naming();
        assert_eq!(naming["ordinary"], ColumnBucketKey::positive(0));
        assert_eq!(naming["kombi2"], ColumnBucketKey::positive(8));
        assert_eq!(naming["metakonkretNot"], ColumnBucketKey::negative(11));
        assert_eq!(bundle.snapshot().bucket_names.len(), 24);
    }

    #[test]
    fn negative_bucket_subtracts_and_disappears() {
        let bundle = bootstrap_column_selection();
        let mut buckets = bundle.new_bucket_map();
        buckets.insert(ColumnBucketKey::positive(0), BTreeSet::from([1, 2, 3]));
        buckets.insert(ColumnBucketKey::negative(0), BTreeSet::from([2]));
        let normalized = bundle.normalize_bucket_map(&buckets);
        assert_eq!(normalized[&ColumnBucketKey::positive(0)], BTreeSet::from([1, 3]));
        assert!(!normalized.contains_key(&ColumnBucketKey::negative(0)));
    }
}
